backend/src/v5/routes/sso/aad/aad.js

```diff
@@ -0,0 +1,165 @@
+/**
+ *  Copyright (C) 2022 3D Repo Ltd
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+const { checkAadConfig, getUserDetailsAndValidateEmail } = require('../../../middleware/dataConverter/inputs/sso/aad');
+const { Router } = require('express');
+const Users = require('../../../processors/users');
+const { getAuthenticationCodeUrl } = require('../../../utils/sso/aad');
+const { respond } = require('../../../utils/responder');
+const { templates } = require('../../../utils/responseCodes');
+const { validateSignUpData } = require('../../../middleware/dataConverter/inputs/users');
+
+const authenticateRedirectUri = 'http://localhost/api/v5/sso/aad/authenticate-post';
+const signupRedirectUri = 'http://localhost/api/v5/sso/aad/signup-post';
+// Frontend uri for signup form
+const sigupUri = '';
+
+const authenticate = (req, res) => {
+	try {
+		const params = { redirectUri: authenticateRedirectUri, state: sigupUri };
+		const authenticationCodeUrl = getAuthenticationCodeUrl(params);
+		res.redirect(authenticationCodeUrl);
+	} catch (err) {
+		/* istanbul ignore next */
+		respond(req, res, err);
+	}
+};
+
+const authenticatePost = (req, res) => {
+	try {
+		res.redirect(req.query.state);
+	} catch (err) {
+		/* istanbul ignore next */
+		respond(req, res, err);
+	}
+};
+
+const signup = (req, res) => {
+	try {
+		const { body } = req;
+		const params = {
+			redirectUri: signupRedirectUri,
+			state: JSON.stringify({
```

Comment:
I think you also need to encodeURI this, as the state becomes a query string

Reply:
I think thats happening authomatically ... is it needed ?
![image](https://user-images.githubusercontent.com/71635617/179808146-83eb6188-e100-4e5a-b597-94552f502ea9.png)

Reply:
if msal do it for you then i guess it's not needed